File: surface/widgets/compose/components/widgets/concepts/llm-safety/GuardrailConfig.kt
package com.clef.surface.widgets.concepts.llmsafety

import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.itemsIndexed
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import kotlinx.coroutines.launch

// --- State machine ---

enum class GuardrailConfigState { Viewing, RuleSelected, Testing, Adding }

sealed class GuardrailConfigEvent {
    data class SelectRule(val id: String) : GuardrailConfigEvent()
    object Test : GuardrailConfigEvent()
    object AddRule : GuardrailConfigEvent()
    object Deselect : GuardrailConfigEvent()
    object TestComplete : GuardrailConfigEvent()
    object Save : GuardrailConfigEvent()
    object Cancel : GuardrailConfigEvent()
}

fun guardrailConfigReduce(state: GuardrailConfigState, event: GuardrailConfigEvent): GuardrailConfigState = when (state) {
    GuardrailConfigState.Viewing -> when (event) {
        is GuardrailConfigEvent.SelectRule -> GuardrailConfigState.RuleSelected
        is GuardrailConfigEvent.Test -> GuardrailConfigState.Testing
        is GuardrailConfigEvent.AddRule -> GuardrailConfigState.Adding
        else -> state
    }
    GuardrailConfigState.RuleSelected -> when (event) {
        is GuardrailConfigEvent.Deselect -> GuardrailConfigState.Viewing
        else -> state
    }
    GuardrailConfigState.Testing -> when (event) {
        is GuardrailConfigEvent.TestComplete -> GuardrailConfigState.Viewing
        else -> state
    }
    GuardrailConfigState.Adding -> when (event) {
        is GuardrailConfigEvent.Save -> GuardrailConfigState.Viewing
        is GuardrailConfigEvent.Cancel -> GuardrailConfigState.Viewing
        else -> state
    }
}

// --- Types ---

data class GuardrailRule(
    val id: String,
    val name: String,
    val description: String,
    val enabled: Boolean,
    val type: String,
    val severity: String
)

data class TestResult(
    val ruleId: String,
    val ruleName: String,
    val triggered: Boolean,
    val severity: String
)

@Composable
fun GuardrailConfig(
    rules: List<GuardrailRule>,
    name: String,
    guardrailType: String,
    modifier: Modifier = Modifier,
    showTest: Boolean = true,
    onRuleToggle: (String, Boolean) -> Unit = { _, _ -> },
    onSeverityChange: (String, String) -> Unit = { _, _ -> },
    onTest: suspend (String) -> List<TestResult> = { emptyList() },
    onAddRule: () -> Unit = {}
) {
    var state by remember { mutableStateOf(GuardrailConfigState.Viewing) }
    var selectedRuleId by remember { mutableStateOf<String?>(null) }
    var testValue by remember { mutableStateOf("") }
    var testResults by remember { mutableStateOf<List<TestResult>>(emptyList()) }
    val scope = rememberCoroutineScope()
    val severityOptions = listOf("block", "warn", "log")

    Column(modifier = modifier.semantics { contentDescription = "Guardrail config: $name" }) {
        // Header
        Row(
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.spacedBy(8.dp),
            modifier = Modifier.padding(bottom = 8.dp)
        ) {
            Text(name, fontWeight = FontWeight.SemiBold, fontSize = 16.sp)
            Surface(tonalElevation = 2.dp, shape = MaterialTheme.shapes.small) {
                Text(guardrailType, fontSize = 11.sp, modifier = Modifier.padding(horizontal = 6.dp, vertical = 2.dp))
            }
        }

        // Rule list
        LazyColumn(modifier = Modifier.weight(1f)) {
            itemsIndexed(rules) { _, rule ->
                val isSelected = selectedRuleId == rule.id
                Card(
                    modifier = Modifier
                        .fillMaxWidth()
                        .padding(vertical = 2.dp)
                        .clickable {
                            selectedRuleId = rule.id
                            state = guardrailConfigReduce(state, GuardrailConfigEvent.SelectRule(rule.id))
                        },
                    colors = CardDefaults.cardColors(
                        containerColor = if (isSelected) MaterialTheme.colorScheme.primaryContainer
                        else MaterialTheme.colorScheme.surface
                    )
                ) {
                    Column(Modifier.padding(8.dp)) {
                        Row(
                            verticalAlignment = Alignment.CenterVertically,
                            horizontalArrangement = Arrangement.spacedBy(8.dp)
                        ) {
                            // Toggle
                            Switch(
                                checked = rule.enabled,
                                onCheckedChange = { onRuleToggle(rule.id, !rule.enabled) },
                                modifier = Modifier.semantics { contentDescription = "Toggle ${rule.name}" }
                            )

                            Column(modifier = Modifier.weight(1f)) {
                                Text(rule.name, fontWeight = FontWeight.Medium, fontSize = 13.sp)
                                Text(rule.description, fontSize = 11.sp, color = MaterialTheme.colorScheme.onSurfaceVariant)
                            }

                            // Type badge
                            Text(rule.type, fontSize = 10.sp, color = MaterialTheme.colorScheme.onSurfaceVariant)
                        }

                        // Severity selector
                        Row(
                            horizontalArrangement = Arrangement.spacedBy(4.dp),
                            modifier = Modifier.padding(top = 4.dp)
                        ) {
                            severityOptions.forEach { sev ->
                                FilterChip(
                                    selected = rule.severity == sev,
                                    onClick = { onSeverityChange(rule.id, sev) },
                                    label = { Text(sev, fontSize = 10.sp) },
                                    modifier = Modifier.height(28.dp)
                                )
                            }
                        }
                    }
                }
            }
        }

        // Add rule button
        TextButton(
            onClick = {
                state = guardrailConfigReduce(state, GuardrailConfigEvent.AddRule)
                onAddRule()
            },
            modifier = Modifier.padding(vertical = 4.dp)
        ) {
            Text("Add Rule", fontSize = 13.sp)
        }

        // Test area
        if (showTest) {
            HorizontalDivider(modifier = Modifier.padding(vertical = 8.dp))
            Text("Rule Tester", fontWeight = FontWeight.SemiBold, fontSize = 14.sp)

            OutlinedTextField(
                value = testValue,
                onValueChange = { testValue = it },
                placeholder = { Text("Enter test input\u2026") },
                modifier = Modifier.fillMaxWidth().padding(vertical = 4.dp),
                minLines = 2
            )

            Button(
                onClick = {
                    state = guardrailConfigReduce(state, GuardrailConfigEvent.Test)
                    scope.launch {
                        try {
                            testResults = onTest(testValue)
                        } finally {
                            state = guardrailConfigReduce(state, GuardrailConfigEvent.TestComplete)
                        }
                    }
                },
                enabled = state != GuardrailConfigState.Testing && testValue.trim().isNotEmpty()
            ) {
                Text(if (state == GuardrailConfigState.Testing) "Testing\u2026" else "Test")
            }

            // Test results
            if (testResults.isNotEmpty()) {
                val triggered = testResults.filter { it.triggered }
                Text(
                    if (triggered.isEmpty()) "All rules passed"
                    else "${triggered.size} rule${if (triggered.size == 1) "" else "s"} triggered",
                    fontSize = 13.sp,
                    modifier = Modifier.padding(top = 4.dp)
                )
                testResults.forEach { result ->
                    Text(
                        "${result.ruleName}: ${if (result.triggered) result.severity else "pass"}",
                        fontSize = 12.sp,
                        color = if (result.triggered) MaterialTheme.colorScheme.error else MaterialTheme.colorScheme.onSurfaceVariant,
                        modifier = Modifier.padding(start = 8.dp)
                    )
                }
            }
        }
    }
}

File: surface/widgets/swiftui/components/widgets/concepts/llm-safety/GuardrailConfig.swift
import SwiftUI

enum GuardrailConfigWidgetState {
    case viewing, ruleSelected, testing, adding
}

enum GuardrailConfigEvent {
    case selectRule, test, addRule, deselect, testComplete, save, cancel
}

func guardrailConfigReduce(state: GuardrailConfigWidgetState, event: GuardrailConfigEvent) -> GuardrailConfigWidgetState {
    switch state {
    case .viewing:
        if event == .selectRule { return .ruleSelected }
        if event == .test { return .testing }
        if event == .addRule { return .adding }
        return state
    case .ruleSelected:
        if event == .deselect { return .viewing }
        return state
    case .testing:
        if event == .testComplete { return .viewing }
        return state
    case .adding:
        if event == .save { return .viewing }
        if event == .cancel { return .viewing }
        return state
    }
}

struct GuardrailRule: Identifiable {
    var id: String
    var name: String
    var description: String
    var enabled: Bool
    var type: String // "input", "output", "both"
    var severity: String // "block", "warn", "log"
}

struct GuardrailTestResult: Identifiable {
    var id: String { ruleId }
    var ruleId: String
    var ruleName: String
    var triggered: Bool
    var severity: String
}

struct GuardrailConfigView: View {
    var rules: [GuardrailRule]
    var name: String
    var guardrailType: String
    var testInput: String = ""
    var showHistory: Bool = true
    var showTest: Bool = true
    var onRuleToggle: ((String, Bool) -> Void)? = nil
    var onSeverityChange: ((String, String) -> Void)? = nil
    var onTest: ((String) -> [GuardrailTestResult])? = nil
    var onAddRule: (() -> Void)? = nil

    @State private var widgetState: GuardrailConfigWidgetState = .viewing
    @State private var selectedRuleId: String? = nil
    @State private var testValue: String = ""
    @State private var testResults: [GuardrailTestResult] = []

    private let severityOptions = ["block", "warn", "log"]

    private func severityColor(_ severity: String) -> Color {
        switch severity {
        case "block": return .red
        case "warn": return .orange
        case "log": return .blue
        default: return .secondary
        }
    }

    var body: some View {
        VStack(alignment: .leading, spacing: 12) {
            // Header
            HStack {
                Text(name)
                    .font(.headline)
                Text(guardrailType)
                    .font(.caption)
                    .padding(.horizontal, 8)
                    .padding(.vertical, 2)
                    .background(Color.secondary.opacity(0.15))
                    .cornerRadius(4)
                    .accessibilityLabel("Type: \(guardrailType)")
            }

            // Rule list
            ForEach(rules) { rule in
                VStack(alignment: .leading, spacing: 6) {
                    HStack {
                        Button(rule.enabled ? "On" : "Off") {
                            onRuleToggle?(rule.id, !rule.enabled)
                        }
                        .accessibilityLabel("Toggle \(rule.name)")
                        .accessibilityValue(rule.enabled ? "enabled" : "disabled")
                        .accessibilityRole(.toggleButton)

                        VStack(alignment: .leading, spacing: 2) {
                            Text(rule.name)
                                .font(.body)
                                .fontWeight(.medium)
                            Text(rule.description)
                                .font(.caption)
                                .foregroundColor(.secondary)
                        }

                        Spacer()

                        Text(rule.type)
                            .font(.caption2)
                            .padding(.horizontal, 6)
                            .padding(.vertical, 2)
                            .background(Color.secondary.opacity(0.1))
                            .cornerRadius(4)
                            .accessibilityLabel("Applies to: \(rule.type)")
                    }

                    // Severity selector
                    HStack(spacing: 4) {
                        ForEach(severityOptions, id: \.self) { sev in
                            Button(sev) {
                                onSeverityChange?(rule.id, sev)
                            }
                            .font(.caption)
                            .padding(.horizontal, 8)
                            .padding(.vertical, 4)
                            .background(rule.severity == sev ? severityColor(sev).opacity(0.2) : Color.clear)
                            .overlay(
                                RoundedRectangle(cornerRadius: 4)
                                    .stroke(rule.severity == sev ? severityColor(sev) : Color.gray.opacity(0.3), lineWidth: 1)
                            )
                            .cornerRadius(4)
                            .accessibilityLabel("Set severity to \(sev)")
                            .accessibilityValue(rule.severity == sev ? "selected" : "")
                        }
                    }

                    if showHistory {
                        Rectangle()
                            .fill(Color.gray.opacity(0.1))
                            .frame(height: 20)
                            .cornerRadius(2)
                            .accessibilityHidden(true)
                    }
                }
                .padding(8)
                .background(
                    RoundedRectangle(cornerRadius: 6)
                        .stroke(selectedRuleId == rule.id ? Color.accentColor : Color.gray.opacity(0.2), lineWidth: selectedRuleId == rule.id ? 2 : 1)
                )
                .onTapGesture {
                    selectedRuleId = rule.id
                    widgetState = guardrailConfigReduce(state: widgetState, event: .selectRule)
                }
                .accessibilityLabel("\(rule.name) \u{2014} \(rule.severity)")
            }

            // Add rule button
            Button("Add Rule") {
                widgetState = guardrailConfigReduce(state: widgetState, event: .addRule)
                onAddRule?()
            }
            .accessibilityLabel("Add new custom rule")

            // Test area
            if showTest {
                VStack(alignment: .leading, spacing: 8) {
                    Text("Rule Tester")
                        .font(.subheadline)
                        .fontWeight(.semibold)

                    TextEditor(text: $testValue)
                        .frame(minHeight: 60)
                        .overlay(
                            RoundedRectangle(cornerRadius: 4)
                                .stroke(Color.gray.opacity(0.3), lineWidth: 1)
                        )
                        .accessibilityLabel("Test input for validating rules")

                    Button(widgetState == .testing ? "Testing..." : "Test") {
                        guard !testValue.trimmingCharacters(in: .whitespaces).isEmpty else { return }
                        widgetState = guardrailConfigReduce(state: widgetState, event: .test)
                        if let results = onTest?(testValue) {
                            testResults = results
                        }
                        widgetState = guardrailConfigReduce(state: widgetState, event: .testComplete)
                    }
                    .disabled(widgetState == .testing || testValue.trimmingCharacters(in: .whitespaces).isEmpty)
                    .accessibilityLabel("Test rules against input")

                    // Test results
                    if !testResults.isEmpty {
                        let triggered = testResults.filter { $0.triggered }
                        Text(triggered.isEmpty ? "All rules passed" : "\(triggered.count) rule\(triggered.count == 1 ? "" : "s") triggered")
                            .font(.caption)
                            .foregroundColor(triggered.isEmpty ? .green : .red)

                        ForEach(testResults) { result in
                            HStack {
                                Text(result.ruleName)
                                    .font(.caption)
                                Spacer()
                                Text(result.triggered ? result.severity : "pass")
                                    .font(.caption)
                                    .foregroundColor(result.triggered ? severityColor(result.severity) : .green)
                            }
                        }
                    }
                }
                .padding(8)
                .overlay(
                    RoundedRectangle(cornerRadius: 6)
                        .stroke(Color.gray.opacity(0.2), lineWidth: 1)
                )
            }
        }
        .accessibilityElement(children: .contain)
        .accessibilityLabel("Guardrail config: \(name)")
        .onAppear {
            testValue = testInput
        }
    }
}
